### ISSUE TITLE
Linear modelling in XGBoost

### ISSUE THREAD

**Issue by RAMitchell** [dmlc/xgboost#3099]
I am planning the following work on improving linear modelling in Xgboost. Comments and feedback welcome. The goal is to make xgboost very competitive with other packages for linear modelling.

- [ ]  Refactor gblinear and add linear updater classes to allow different linear model implementations
- [ ]  Normalise l1 and l2 penalties to be consistent with number of training examples (same as sklear/glmnet)
- [ ]  Faster more accurate training with new coordinate descent algorithm

- [ ]  Early stopping based on user specified tolerance parameter. Early stopping based on eval metric is not accurate enough and doesn't consider regularisation

- [ ] More unit tests. Add tests for Poisson/Gamma (I suspect these will not work with current shotgun implementation)

- [ ] AVX/GPU acceleration

- [ ]  Add Newton method or bfgs based solver

**Comment by khotilov** [comment]
I agree that the linear booster is an underdog that needs some good care. And GPU might really accelerate things here, especially if some of the Nvidia's matrix algebra could be used.

BTW, I have a branch that I started working on in Aug https://github.com/dmlc/xgboost/compare/master...khotilov:gblinear_dev?expand=1
that has a penalty normalization options and adds a top-N components stagewise algorithm. I never got time to finish the work to a clean state (add the updater interface, etc.). Do you mind me merging that later in the evening after doing some minimal reconciliation (for the predictor interface)?

And I'm not really sure what that early stopping issue is. The tolerance could be set in the wrapper interface.

**Comment by RAMitchell** [comment]
I'm going to PR my work in the next couple of days. Let's just PR both and see what makes sense from there.

I want to do early stopping based on the delta weight. I.e. if the largest weight update in an iteration falls below tolerance. Sklearn uses a more sophisticated stopping criterea based on the duality gap that could also be an option. I'm open to ideas how to do that. My current thought was to add a return value to the DoBoost  c api function indicating if the algorithm has converged, then the python/r code can stop iterating.
